File: modules/module-provider/src/main/scala/io/constellation/provider/sdk/CanaryCoordinator.scala
package io.constellation.provider.sdk

import cats.effect.IO
import cats.implicits.*

// ===== Canary Result =====

enum CanaryResult:
  case Promoted
  case RolledBack(reason: String)
  case PartialFailure(promoted: List[String], failed: List[String])

// ===== CanaryCoordinator =====

/** Coordinates sequential canary rollout across instances.
  *
  * For each instance: replace modules → wait observation window → check health.
  * If unhealthy: rollback all previously upgraded instances to old modules.
  */
class CanaryCoordinator(
    connections: Map[String, InstanceConnection],
    config: CanaryConfig
) {

  /** Roll out new modules sequentially across all instances.
    *
    * Returns Promoted if all instances pass health check,
    * RolledBack if any instance fails (with rollback of previously upgraded instances).
    */
  def rollout(
      oldModules: List[ModuleDefinition],
      newModules: List[ModuleDefinition]
  ): IO[CanaryResult] = {
    val instances = connections.toList.sortBy(_._1)

    if instances.isEmpty then IO.pure(CanaryResult.Promoted)
    else rolloutSequential(instances, oldModules, newModules, upgraded = List.empty)
  }

  private def rolloutSequential(
      remaining: List[(String, InstanceConnection)],
      oldModules: List[ModuleDefinition],
      newModules: List[ModuleDefinition],
      upgraded: List[(String, InstanceConnection)]
  ): IO[CanaryResult] =
    remaining match {
      case Nil =>
        IO.pure(CanaryResult.Promoted)

      case (instanceId, conn) :: rest =>
        for {
          _       <- conn.replaceModules(newModules)
          _       <- IO.sleep(config.observationWindow)
          healthy <- conn.isHealthy
          result <- if healthy then
            rolloutSequential(rest, oldModules, newModules, upgraded :+ (instanceId -> conn))
          else {
            // Rollback all previously upgraded instances
            val allToRollback = upgraded :+ (instanceId -> conn)
            allToRollback.traverse_ { case (_, c) => c.replaceModules(oldModules) }.as(
              CanaryResult.RolledBack(s"Instance $instanceId failed health check after canary deployment")
            )
          }
        } yield result
    }
}

File: sdks/typescript/src/provider/canary-coordinator.ts
/**
 * Coordinates sequential canary rollout across instances.
 *
 * For each instance: replace modules -> wait observation window -> check health.
 * If unhealthy: rollback all previously upgraded instances to old modules.
 *
 * Mirrors the Scala CanaryCoordinator class.
 */

import type { CanaryConfig } from "../types/config.js";
import type { CanaryResult } from "../types/canary.js";
import type { ModuleDefinition } from "../types/module-definition.js";
import type { InstanceConnection } from "./instance-connection.js";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class CanaryCoordinator {
  private readonly connections: Map<string, InstanceConnection>;
  private readonly config: CanaryConfig;

  constructor(connections: Map<string, InstanceConnection>, config: CanaryConfig) {
    this.connections = connections;
    this.config = config;
  }

  /** Roll out new modules sequentially across all instances. */
  async rollout(
    oldModules: ModuleDefinition[],
    newModules: ModuleDefinition[],
  ): Promise<CanaryResult> {
    const instances = [...this.connections.entries()].sort(([a], [b]) => a.localeCompare(b));

    if (instances.length === 0) {
      return { status: "Promoted" };
    }

    return this.rolloutSequential(instances, oldModules, newModules, []);
  }

  private async rolloutSequential(
    remaining: Array<[string, InstanceConnection]>,
    oldModules: ModuleDefinition[],
    newModules: ModuleDefinition[],
    upgraded: Array<[string, InstanceConnection]>,
  ): Promise<CanaryResult> {
    if (remaining.length === 0) {
      return { status: "Promoted" };
    }

    const [[instanceId, conn], ...rest] = remaining;

    conn.replaceModules(newModules);
    await sleep(this.config.observationWindowMs);
    const healthy = conn.isHealthy;

    if (healthy) {
      return this.rolloutSequential(rest, oldModules, newModules, [
        ...upgraded,
        [instanceId, conn],
      ]);
    } else {
      // Rollback all previously upgraded instances + this one
      const allToRollback = [...upgraded, [instanceId, conn]] as Array<
        [string, InstanceConnection]
      >;
      for (const [, c] of allToRollback) {
        c.replaceModules(oldModules);
      }
      return {
        status: "RolledBack",
        reason: `Instance ${instanceId} failed health check after canary deployment`,
      };
    }
  }
}
